uber/site_sections/preregistration.py

```diff
@@ -562,10 +562,25 @@ def invalid_badge(self, session, id, message=''):
     def not_found(self, id, message):
         return {'id': id, 'message': message}
 
-    def invalidate(self, session, id):
+    def abandon_badge(self, session, id):
         attendee = session.attendee(id)
-        attendee.badge_status = c.INVALID_STATUS
-        raise HTTPRedirect('invalid_badge?id={}&message={}', attendee.id, 'Sorry you can\'t make it! We hope to see you next year!')
+        success_message = "Sorry you can't make it! We hope to see you next year!"
+        failure_message = "You cannot abandon your badge because you are the leader of a group."
+
+        # if attendee is part of a group, we must delete attendee and remove from the group
+        if attendee.group:
+            if attendee.group.leader != attendee:
+                # remove this attendee from the group and delete the attendee
+                session.assign_badges(attendee.group, attendee.group.badges + 1, new_badge_type=attendee.badge_type, new_ribbon_type=attendee.ribbon, registered=attendee.registered, paid=attendee.paid)
```

Comment:
What you did here was totally fine, but this is now the third place in the code where we're doing a "group badge unassign" function -- the other two places are on the group management page and on the admin side. It's probably getting about time where we want a separate function for that operation. Out of scope for this PR, though!

Reply:
This occurred to me, but I was not sure where this new extracted function should be defined. It's interesting to hear that this doesn't necessarily belong in this PR, as I would normally consider extracting code into a separate function as soon as I duplicate it.